compiler/src/dotty/tools/dotc/core/GadtConstraint.scala

```diff
@@ -11,38 +11,95 @@ import collection.mutable
 import printing._
 
 import scala.annotation.internal.sharable
+import Denotations.{Denotation, SingleDenotation}
+import SymDenotations.NoDenotation
+
+/** Types that represent a path. Can either be a TermRef or a SkolemType. */
+type PathType = TermRef | SkolemType
 
 /** Represents GADT constraints currently in scope */
 sealed abstract class GadtConstraint extends Showable {
   /** Immediate bounds of `sym`. Does not contain lower/upper symbols (see [[fullBounds]]). */
   def bounds(sym: Symbol)(using Context): TypeBounds | Null
 
+  /** Immediate bounds of a path-dependent type.
+    * This variant of bounds will ONLY try to retrieve path-dependent GADT bounds. */
+  def bounds(path: PathType, sym: Symbol)(using Context): TypeBounds | Null
+  def bounds(tp: TypeRef)(using Context): TypeBounds | Null
+
   /** Full bounds of `sym`, including TypeRefs to other lower/upper symbols.
    *
    * @note this performs subtype checks between ordered symbols.
    *       Using this in isSubType can lead to infinite recursion. Consider `bounds` instead.
    */
   def fullBounds(sym: Symbol)(using Context): TypeBounds | Null
 
+  def fullBounds(path: PathType, sym: Symbol)(using Context): TypeBounds | Null
+  def fullBounds(tp: TypeRef)(using Context): TypeBounds | Null
+
   /** Is `sym1` ordered to be less than `sym2`? */
   def isLess(sym1: Symbol, sym2: Symbol)(using Context): Boolean
 
+  /** Is `tp1` ordered to be less than `tp2`? */
+  def isLess(tp1: NamedType, tp2: NamedType)(using Context): Boolean
+
   /** Add symbols to constraint, correctly handling inter-dependencies.
    *
    * @see [[ConstraintHandling.addToConstraint]]
    */
   def addToConstraint(syms: List[Symbol])(using Context): Boolean
   def addToConstraint(sym: Symbol)(using Context): Boolean = addToConstraint(sym :: Nil)
 
+  /** Add path to constraint, registering all its abstract type members. */
+  def addToConstraint(path: PathType)(using Context): Boolean
+
   /** Further constrain a symbol already present in the constraint. */
   def addBound(sym: Symbol, bound: Type, isUpper: Boolean)(using Context): Boolean
 
+  /** Further constrain a path-dependent type already present in the constraint. */
+  def addBound(p: PathType, sym: Symbol, bound: Type, isUpper: Boolean)(using Context): Boolean
+
+  /** Record the equality between two singleton types. */
+  def addEquality(p: PathType, q: PathType)(using Context): Unit
+
+  /** Check whether two singleton types are equivalent. */
+  def isEquivalent(p: PathType, q: PathType): Boolean
+
+  /** Query the representative member of a singleton type. */
+  def reprOf(p: PathType): PathType | Null
+
+  /** Scrutinee path of the current pattern matching. */
+  def scrutineePath: TermRef | Null
```

Comment:
Could you explain why do we need `scrutineePath`, and why we need to reset it?

Reply:
I added detailed documentation in the docstring of `constrainTypeMembers` ([commit1](https://github.com/lampepfl/dotty/pull/14754/commits/e043fdd14979b0503794a378e94b45c8cf0c3e9c), [commit2](https://github.com/lampepfl/dotty/pull/14754/commits/7d5046ff97d807e7d5e1d28e44250c2b6fbc5fb7)). Does it look good?